// @layer0/cli/commands/env/pull.js
"use strict";

const {
  EOL
} = require('os');

const path = require('path');

const fse = require('fs-extra');

const {
  confirmOrExit
} = require('../../utils/prompts');

const authenticate = require('../../prompts/authenticate');

const Layer0PackageJson = require('../../utils/Layer0PackageJson');

exports.command = 'pull <path-to-env-file>';
exports.describe = 'Copy non-secret environment variables from an environment to a local .env file';
exports.builder = {
  site: {
    type: 'string',
    alias: 's',
    describe: 'Slug of the site to pull variables from. Using package.json name property if omitted.'
  },
  team: {
    type: 'string',
    alias: 't',
    describe: 'The name of the team under which the site belongs. Using private space if omitted.'
  },
  environment: {
    type: 'string',
    alias: 'e',
    describe: 'Environment to pull variables from. Uses default environment if omitted.',
    default: 'default'
  }
};

exports.handler = async ({
  context,
  pathToEnvFile,
  site,
  team: teamSlug,
  environment
}) => {
  const {
    logger
  } = context;
  const absolutePathToEnvFile = path.resolve(process.cwd(), pathToEnvFile);
  await authenticate(context);

  if (fse.existsSync(absolutePathToEnvFile)) {
    logger.warn(`File ${absolutePathToEnvFile} already exists and will be overwritten`);
    await confirmOrExit('Do you want to continue?', {
      initial: false
    });
  }

  const packageJson = Layer0PackageJson.loadPackageJson('.');
  const siteSlug = site || packageJson.name;
  logger.title(EOL + '📋 Pulling environment variables from:');
  const environmentDetails = [`> team=${teamSlug || 'Private space'}`, `> site=${siteSlug}`, `> environment=${environment}`].filter(Boolean); // Removes falsy values

  logger.info(environmentDetails.join(EOL));
  logger.title(`${EOL}Into file: ${absolutePathToEnvFile}`);
  const environmentVariables = await context.api.getEnvironmentVariables({
    siteSlug,
    teamSlug,
    environment
  });
  let fileContent = '# Environments variables pulled from:' + EOL + environmentDetails.map(detail => `# ${detail}`).join(EOL) + EOL + EOL;

  for (const {
    key,
    secret,
    value
  } of environmentVariables) {
    if (secret) {
      logger.warn(`Warning: Environment variable "${key}" has a secret value. You will need to edit it manually.`);
      fileContent += '# This value is a secret and needs to be edited manually' + EOL;
    }

    fileContent += (secret ? `#${key}=<SECRET VALUE>` : `${key}=${value}`) + EOL + EOL;
  }

  await fse.writeFile(absolutePathToEnvFile, fileContent);
};
